gto-support-api/src/main/java/org/ccci/gto/android/common/api/MultipartFormOutputWriter.java

```diff
@@ -0,0 +1,100 @@
+package org.ccci.gto.android.common.api;
+
+import android.support.annotation.NonNull;
+
+import com.google.common.io.Closer;
+
+import org.ccci.gto.android.common.api.AbstractApi.Request.MediaType;
+import org.ccci.gto.android.common.api.AbstractApi.Request.Parameter;
+import org.ccci.gto.android.common.util.IOUtils;
+
+import java.io.Closeable;
+import java.io.DataOutputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+final class MultipartFormOutputWriter implements Closeable {
+    private static final String PREFIX = "--";
+    private static final String NEWLINE = "\r\n";
+
+    @NonNull
+    private final DataOutputStream mOut;
+    @NonNull
+    private final String mBoundary;
+
+    MultipartFormOutputWriter(@NonNull final OutputStream os, @NonNull final String boundary) {
+        mOut = new DataOutputStream(os);
+        mBoundary = boundary;
+    }
+
+    void writeParameter(@NonNull final Parameter param) throws IOException {
+        if (param.mFile != null) {
+            writeFileParameter(param);
+        } else {
+            writeSimpleParameter(param);
+        }
+    }
+
+    private void writeSimpleParameter(@NonNull final Parameter param) throws IOException {
+        // write boundary
+        mOut.writeBytes(PREFIX);
+        mOut.writeBytes(mBoundary);
+        mOut.writeBytes(NEWLINE);
+
+        // write content header
+        mOut.writeBytes("Content-Disposition: form-data; name=\"" + param.mName + "\"");
+        mOut.writeBytes(NEWLINE);
+        mOut.writeBytes(NEWLINE);
+
+        // write content
+        mOut.writeBytes(param.mValue != null ? param.mValue : "");
+        mOut.writeBytes(NEWLINE);
+    }
+
+    private void writeFileParameter(@NonNull final Parameter param) throws IOException {
+        // write boundary
+        mOut.writeBytes(PREFIX);
+        mOut.writeBytes(mBoundary);
+        mOut.writeBytes(NEWLINE);
+
+        // write content header
+        final String fileName =
+                param.mFileName != null ? param.mFileName : param.mFile != null ? param.mFile.getName() : "";
```

Comment:
I really wish I could use some of perl's syntactic sugar here

`final String fileName = param.mFileName // param.mFile.getName() // "";`